![New York City schoolbus](schoolbus.jpg)

Photo by [Jannis Lucas](https://unsplash.com/@jannis_lucas) on [Unsplash](https://unsplash.com).
<br>

Every year, American high school students take SATs, which are standardized tests intended to measure literacy, numeracy, and writing skills. There are three sections - reading, math, and writing, each with a **maximum score of 800 points**. These tests are extremely important for students and colleges, as they play a pivotal role in the admissions process.

Analyzing the performance of schools is important for a variety of stakeholders, including policy and education professionals, researchers, government, and even parents considering which school their children should attend. 

You have been provided with a dataset called `schools.csv`, which is previewed below.

You have been tasked with answering three key questions about New York City (NYC) public school SAT performance.

In [2]:
# Re-run this cell 
import pandas as pd

# Read in the data
schools = pd.read_csv("schools.csv")

# Preview the first 5 data
schools.head(5)

school_name    borough building_code  \
0  New Explorations into Science, Technology and ...  Manhattan          M022   
1                               Essex Street Academy  Manhattan          M445   
2                       Lower Manhattan Arts Academy  Manhattan          M445   
3    High School for Dual Language and Asian Studies  Manhattan          M445   
4      Henry Street School for International Studies  Manhattan          M056   

   average_math  average_reading  average_writing  percent_tested  
0           657              601              601             NaN  
1           395              411              387            78.9  
2           418              428              415            65.1  
3           613              453              463            95.9  
4           410              406              381            59.7

<h2>Understanding the Data</h2>

<h3>1. Check data information</h3>
Check data types with schools.info(). This is crucial to avoid issues during analysis, especially when dealing with mixed data types (numbers, text) that could lead to unexpected results.


There are 375 entries or rows with 7 columns.

In [3]:
schools.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 375 entries, 0 to 374
Data columns (total 7 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   school_name      375 non-null    object 
 1   borough          375 non-null    object 
 2   building_code    375 non-null    object 
 3   average_math     375 non-null    int64  
 4   average_reading  375 non-null    int64  
 5   average_writing  375 non-null    int64  
 6   percent_tested   355 non-null    float64
dtypes: float64(1), int64(3), object(3)
memory usage: 20.6+ KB


Used .describe() to understand the spread and central tendency of numerical features. This helps identify potential outliers and assess the overall data distribution.

We can see from the results that the ‘percent_tested’ column has 20 missing values. This means there is data for only 355 out of 375 schools.

In [12]:
schools.describe()

average_math  average_reading  average_writing  percent_tested
count    375.000000       375.000000       375.000000      355.000000
mean     432.944000       424.504000       418.458667       64.976338
std       71.952373        61.881069        64.548599       18.747634
min      317.000000       302.000000       284.000000       18.500000
25%      386.000000       386.000000       382.000000       50.950000
50%      415.000000       413.000000       403.000000       64.800000
75%      458.500000       445.000000       437.500000       79.600000
max      754.000000       697.000000       693.000000      100.000000

Out of 375 entries, they are 20 rows or schools with empty "percent_tested" data which is represents approximatelly 5% of the data.

<h3>2. Finding Missing Values</h3>
Out of 375 entries, they are 20 rows or schools with empty “percent_tested” data which is represents approximatelly 5% of the data.

We need to find out how many schools have missing values in the ‘percent_tested’ column. This will tell us which schools didn’t provide data on the percentage of students who took the tests.

In [10]:
schools.isna().sum()

school_name         0
borough             0
building_code       0
average_math        0
average_reading     0
average_writing     0
percent_tested     20
dtype: int64

20 rows with missing data in percent_tested columns

In [23]:
print(schools[schools.isna().any(axis=1)])

                                           school_name  ... percent_tested
0    New Explorations into Science, Technology and ...  ...            NaN
19   Urban Assembly School of Business for Young Women  ...            NaN
38               High School of Hospitality Management  ...            NaN
57      Frederick Douglass Academy II Secondary School  ...            NaN
76              Jacqueline Kennedy Onassis High School  ...            NaN
95                             Tottenville High School  ...            NaN
114                    Bronx International High School  ...            NaN
133                  West Bronx Academy for the Future  ...            NaN
152              Bronx High School for the Visual Arts  ...            NaN
171              Kingsbridge International High School  ...            NaN
190  Archimedes Academy for Math, Science, and Tech...  ...            NaN
209         Young Women's Leadership School in Astoria  ...            NaN
228  High School for Publ

20 Schools with missing data for ‘percent_tested’ columns

While there are identified missing values in the ‘percent_tested’ column, I won’t be removing these rows as this data won’t be used in the current analysis. However, it’s important to note that these missing values could introduce bias or anomalies if not addressed in future analyses that might utilize this column.

<h3>2. Finding Duplicate Values</h3>
Checks for possible duplicated rows in the schools DataFrame.

In [19]:
duplicates = schools.duplicated()
print(schools[duplicates])

Empty DataFrame
Columns: [school_name, borough, building_code, average_math, average_reading, average_writing, percent_tested]
Index: []


No duplicate rows detected

## Question No.1
## Which NYC schools have the best math results?
* The best math results are at least 80% of the *maximum possible score of 800* for math.
* Save your results in a pandas DataFrame called best_math_schools, including "school_name" and "average_math" columns, sorted by "average_math" in descending order.

In [27]:
#result is 80% of max score
max_score = 800
treshold = max_score * .8
print(treshold)

best_math_schools = schools[schools['average_math'] >= treshold][['school_name', 'average_math']].sort_values('average_math', ascending=False)

best_math_schools

640.0


school_name  average_math
88                              Stuyvesant High School           754
170                       Bronx High School of Science           714
93                 Staten Island Technical High School           711
365  Queens High School for the Sciences at York Co...           701
68   High School for Mathematics, Science, and Engi...           683
280                     Brooklyn Technical High School           682
333                        Townsend Harris High School           680
174  High School of American Studies at Lehman College           669
0    New Explorations into Science, Technology and ...           657
45                       Eleanor Roosevelt High School           641

## Question No.2
## What are the top 10 performing schools based on the combined SAT scores?
* Save your results as a pandas DataFrame called top_10_schools containing the "school_name" and a new column named "total_SAT", with results ordered by "total_SAT" in descending order.

In [48]:
schools["total_SAT"] = schools["average_math"] + schools["average_reading"] + schools["average_writing"] 
top_10_schools = schools[['school_name', 'total_SAT']].sort_values('total_SAT', ascending=False).head(10).reset_index(drop=True)

top_10_schools

school_name  total_SAT
0                             Stuyvesant High School       2144
1                       Bronx High School of Science       2041
2                Staten Island Technical High School       2041
3  High School of American Studies at Lehman College       2013
4                        Townsend Harris High School       1981
5  Queens High School for the Sciences at York Co...       1947
6                     Bard High School Early College       1914
7                     Brooklyn Technical High School       1896
8                      Eleanor Roosevelt High School       1889
9  High School for Mathematics, Science, and Engi...       1889

## Question No.3
## Which single borough has the largest standard deviation in the combined SAT score?

* Save your results as a pandas DataFrame called largest_std_dev.
* The DataFrame should contain one row, with:

    * "borough" - the name of the NYC borough with the largest standard deviation of "total_SAT".
    * "num_schools" - the number of schools in the borough.
    * "average_SAT" - the mean of "total_SAT".
    * "std_SAT" - the standard deviation of "total_SAT".
* Round all numeric values to two decimal places.



In [53]:
boroughs = schools.groupby('borough')['total_SAT'].agg(['count', 'mean', 'std']).round(2)

boroughs


count     mean     std
borough                              
Bronx             98  1202.72  150.39
Brooklyn         109  1230.26  154.87
Manhattan         89  1340.13  230.29
Queens            69  1345.48  195.25
Staten Island     10  1439.00  222.30

Summary of the total SAT score distribution for each borough

This provides valuable insights into how total SAT scores vary across different boroughs in the dataset.

In [55]:
value = boroughs['std'].max()
largest_std_dev = boroughs[boroughs['std'] == value]
largest_std_dev = largest_std_dev.rename(columns={"count":"num_schools", "mean":"average_SAT", "std":"std_SAT"})
largest_std_dev.reset_index(inplace=True)

largest_std_dev

borough  num_schools  average_SAT  std_SAT
0  Manhattan           89      1340.13   230.29

Manhattan is the borough with the largest standard deviation